trustmanager/keyfilestore.go

```diff
@@ -106,54 +110,97 @@ func (s *KeyMemoryStore) ListKeys() []string {
 	return listKeys(s)
 }
 
-func addKey(s LimitedFileStore, name string, privKey data.PrivateKey) error {
+// RemoveKey removes the key from the keystore
+func (s *KeyMemoryStore) RemoveKey(name string) error {
+	return removeKey(s, name)
+}
+
+
+func addKey(s LimitedFileStore, passphraseRetriever PassphraseRetriever, name, alias string, privKey data.PrivateKey) error {
 	pemPrivKey, err := KeyToPEM(privKey)
 	if err != nil {
 		return err
 	}
 
-	return s.Add(name, pemPrivKey)
-}
+	attempts := 0
+	passphrase := ""
+	giveup := false
+	for {
+		passphrase, giveup, err = passphraseRetriever(name, alias, true, attempts)
+		if err != nil {
+			attempts++
+			continue
+		} else if giveup {
+			return err
+		} else {
+			break
+		}
+	}
 
-func getKey(s LimitedFileStore, name string) (data.PrivateKey, error) {
-	keyBytes, err := s.Get(name)
-	if err != nil {
-		return nil, err
+	if passphrase != "" {
+		pemPrivKey, err = EncryptPrivateKey(privKey, passphrase)
+		if err != nil {
+			return err
+		}
 	}
 
-	// Convert PEM encoded bytes back to a PrivateKey
-	privKey, err := ParsePEMPrivateKey(keyBytes, "")
+	err = s.Add(name + "." + aliasExtension, []byte(alias))
 	if err != nil {
-		return nil, err
+		return err
 	}
-
-	return privKey, nil
+	return s.Add(name, pemPrivKey)
 }
 
-func addEncryptedKey(s LimitedFileStore, name string, privKey data.PrivateKey, passphrase string) error {
-	encryptedPrivKey, err := EncryptPrivateKey(privKey, passphrase)
+
+func getKeyAlias(s LimitedFileStore, name string) (string, error) {
+	keyAlias, err := s.Get(name + "." + aliasExtension)
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	return s.Add(name, encryptedPrivKey)
+	return string(keyAlias), nil
 }
 
-func getDecryptedKey(s LimitedFileStore, name string, passphrase string) (data.PrivateKey, error) {
+// GetKey returns the PrivateKey given a KeyID
+func getKey(s LimitedFileStore, passphraseRetriever PassphraseRetriever, name string) (data.PrivateKey, error) {
 	keyBytes, err := s.Get(name)
 	if err != nil {
 		return nil, err
 	}
 
-	// Gets an unencrypted PrivateKey.
-	privKey, err := ParsePEMPrivateKey(keyBytes, passphrase)
+	keyAlias, err := getKeyAlias(s, name)
 	if err != nil {
 		return nil, err
 	}
 
+
+	// See if the key is encrypted. If its encrypted we'll fail to parse the private key
+	privKey, err := ParsePEMPrivateKey(keyBytes, "")
+	if err != nil {
+		// We need to decrypt the key, lets get a passphrase
+		attempts := 0
+		for {
+			passphrase, giveup, err := passphraseRetriever(name, string(keyAlias), false, attempts)
+			// Check if the passphrase retriever got an error or if it is telling us to give up
+			if giveup || err != nil {
+				return nil, err
+			}
+
+			// Try to convert PEM encoded bytes back to a PrivateKey using the passphrase
+			privKey, err = ParsePEMPrivateKey(keyBytes, passphrase)
+			if err == nil {
```

Comment:
Normally I would suggest checking what type of error is returned so we can distinguish decryption failure from other types of errors, but I've noticed that the PEM functions can return unexpected errors sometimes when you pass the wrong passphrase, so I think what this code is doing right now is actually best.